Add support for ActionSelector extern to PSA/eBPF backend (#2)
tatry/p4c: backends/ebpf/psa/README.md

**Note:** As of April 2022, support for ActionProfile extern in `psabpf-ctl` CLI/API is not implemented yet. As a workaround
you can use `psabpf-ctl table` command for this extern.

### ActionSelector

> [PSA specification](https://p4.org/p4-spec/docs/PSA-v1.1.0.html#sec-action-selector) is a table implementation similar to
> an ActionProfile, but extends its functionality with support for group of action. If a table entry contains member reference,
> ActionSelector behaves in the same way as ActionProfile. For group references compiler generates additional maps and code
> in order to resolve group reference into member reference. One of additional maps is a map (hash map of maps) that maps
> group reference into inner map (array map) which describes a group. Such map (might be created at runtime by `psabpf-ctl`)
> contains number of members in group in first element, other contain member references. To choose a member from a group,
> checksum is calculated from keys with match type `selector`. Next, obtained member reference from group map is used to
> get action and execute it.

Reviewer: ```suggestion
[PSA specification](https://p4.org/p4-spec/docs/PSA-v1.1.0.html#sec-action-selector) is a table implementation similar to
an ActionProfile, but extends its functionality with support for a group of actions. If a table entry contains a member reference,
the ActionSelector behaves in the same way as an ActionProfile. For group references, the PSA-eBPF compiler generates additional BPF maps. One of additional BPF maps (hash map of maps) maps a group reference ID to an inner map (BPF array map) that contains a group of entries. The inner map (might be created at runtime by `psabpf-ctl`) stores a number of all members in a group as the first element of the inner map. The rest of entries contains members of the ActionSelector group. To choose a member from a group, checksum is calculated from all `selector` match keys. Next, an obtained member from the group map is used to get and execute an action.
```

> Second created map by compiler is a map containing an action for empty group. For ActionSelector there are two fields
> stored in a table that uses given ActionSelector instance, one is reference, second is marker whether reference points
> to group or member.

Reviewer: Próbowałem to przeredagować jak poniżej:

```suggestion
A second created map contains an action for an empty group. For the ActionSelector, there are two fields stored in a table that uses given ActionSelector instance, one is reference, second is marker whether reference points to group or member.
```

ale nie mam pojecia co miałeś na myśli w tym zdaniu: `there are two fields stored in a table that uses given ActionSelector instance, one is reference, second is marker whether reference points to group or member.`

                } else {
                    /* Not found, probably bug. Skip further execution of the extern. */
                    bpf_trace_message("ActionSelector: Entry with action reference was not found, dropping packet. Bug?\n");
                    return TC_ACT_SHOT;
                }
            } else {
                bpf_trace_message("ActionSelector: empty group, going to default action\n");
                as_group_state = 1;
            }
        } else {
            bpf_trace_message("ActionSelector: entry with number of elements not found, dropping packet. Bug?\n");
            return TC_ACT_SHOT;
        }
    } else {
        bpf_trace_message("ActionSelector: group map was not found, dropping packet. Bug?\n");
        return TC_ACT_SHOT;
    }
}
if (as_group_state == 0) {
    bpf_trace_message("ActionSelector: member reference %u\n", as_action_ref);
    as_value = BPF_MAP_LOOKUP_ELEM(ingress_as_actions, &as_action_ref);         // (6)
} else if (as_group_state == 1) {
    bpf_trace_message("ActionSelector: empty group, executing default group action\n");
    as_value = BPF_MAP_LOOKUP_ELEM(ingress_as_defaultActionGroup, &ebpf_zero);  // (7)
}
```

> Description of marked lines:
> 1. Detect if reference is group reference. When field `_is_group_ref` is non-zero, reference is assumed to be a group
>    reference.
> 2. Read first entry in a group. This gives number of members in a group.
> 3. From calculated hash some least significant bits are taken into account. Number of this bits are equal to last
>    parameter of constructor of `ActionSelector`.
> 4. Number of members in a group is known (first entry in a table) and one of them must be chosen. Based on hash value,
>    result is an action id in a group. Valid value of action id in a group is in the set {1, 2, ... number of members}
> 5. This lookup is necessary to translate action id in a group into member reference.
> 6. When member reference is found (for a group; when member reference was read from table, it is used here) action data
>    is read from `_actions` map.
> 7. For an empty group (without members) action data is read from `_defaultActionGroup` table.

Reviewer: ```suggestion
Description of marked lines:
1. Detect if a reference is a group reference. When the `_is_group_ref` field is non-zero, the reference is assumed to be a group reference.
2. Read a first entry in a group. This gives the number of members in a group.
3. A number of least significant bits from calculated hash are taken into account. The number of bits are equal to the last parameter of the ActionSelector constructor.
4. The number of members in a group is known (the first entry in a table) and one of them must be dynamically selected.  An action ID in a group is chosen based on the calculated hash value. A valid value of an action ID in a group is within the following range: {1, 2, ... number of members}. 
5. This lookup is necessary to translate the action ID in a group into a member reference.
6. When a member reference is found, action data is read from the `_actions` map.
7. For an empty group (without members), action data is read from the `_defaultActionGroup` table.
```

> To manage ActionSelector instance (do not confuse with table that uses this implementation), you can use 
> `psabpf-ctl action-selector` command or C API from psabpf.

Reviewer: ```suggestion
To manage the ActionSelector instance (do not confuse with a table that uses this implementation), you can use 
`psabpf-ctl action-selector` command or C API from psabpf.
```
